File: C++/Tree/2583. Kth Largest Sum in a Binary Tree.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int kthLargestLevelSum(TreeNode* root, int k) {
        if (!root) return -1;

        queue<TreeNode*> q;
        q.push(root);
        vector<int> level_sums;

        while (!q.empty()) {
            int nodes_len = q.size();
            int sums = 0;

            for (int i = 0; i < nodes_len; ++i) {
                TreeNode* node = q.front();
                q.pop();
                sums += node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }

            level_sums.push_back(sums);
        }

        sort(level_sums.begin(), level_sums.end(), greater<int>());

        return k <= level_sums.size() ? level_sums[k - 1] : -1;
    }
};

int main() {
    // Creating the binary tree
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(8);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(4);
    root->right->left = new TreeNode(7);
    root->right->right = new TreeNode(9);

    // Solution instance
    Solution solution;
    int k = 2;

    // Output the result
    cout << "K-th largest level sum: " << solution.kthLargestLevelSum(root, k) << endl;

    // Free memory
    delete root->left->left;
    delete root->left->right;
    delete root->right->left;
    delete root->right->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

File: Python/tree/2583. Kth Largest Sum in a Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Example usage:
root = TreeNode(5)  # Root node
root.left = TreeNode(3)  # Left child of root
root.right = TreeNode(8)  # Right child of root
root.left.left = TreeNode(1)  # Left child of root.left
root.left.right = TreeNode(4)  # Right child of root.left
root.right.left = TreeNode(7)  # Left child of root.right
root.right.right = TreeNode(9)  # Right child of root.right



class Solution:
    def kthLargestLevelSum(self, root, k: int) -> int:
        if not root:
            return root
        q = [root]
        level_sums = []
        while q:
            nodes_len = len(q)
            sums = 0
            for _ in range(nodes_len):
                node = q.pop(0)
                print(node.val)
                sums += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            print("--")

            level_sums.append(sums)
        print(level_sums)
        return sorted(level_sums)[-k] if k <= len(level_sums) else -1

# Now you can use the kthLargestLevelSum function to find the k-th largest level sum.
solution = Solution()
k = 2
print("K-th largest level sum:", solution.kthLargestLevelSum(root, k))